### DISCUSSION TITLE
Missing information about what CRC changes in the machine

### DISCUSSION THREAD

**Discussion by didib** [crc-org/crc#3211] [General]
The [getting started](https://crc.dev/crc/) guide does not mention anywhere e.g. "/etc/hosts" or "resolv.conf". I'd like, as a user, if I consider using CRC, to know beforehand, what it's going to touch on my laptop - both as root/sudo (/etc/*) and in my home directory.

Perhaps better not have this hard-coded in docs/getting-started but have a 'crc' sub-command to show it.

What I found out so far (by having /etc/ maintained in git and looking at my home directory, didn't check elsewhere nor the code):

### crc setup:
- Creates /etc/NetworkManager/conf.d/crc-nm-dnsmasq.conf
- Creates /etc/NetworkManager/dnsmasq.d/crc.conf
- Adds my account to the group 'libvirt'
- Creates /etc/libvirt/qemu/networks/autostart/crc.xml
- Creates /etc/libvirt/qemu/networks/crc.xml
- Changes /etc/resolv.conf (and see also #1663 which I ran into)

### crc start:
- Changes /etc/hosts
- Creates /etc/libvirt/qemu/crc.xml
- Creates /etc/libvirt/storage/crc.xml

### Both of them (didn't check which - likely setup):
- Create $HOME/.config/systemd/user/crc* - user systemd units
- Create $HOME/.crc - crc itself - perhaps should be made configurable somehow

**Comment by gbraad** [comment]
Added the trivial priority as our focus is on merging our development branch to master. this means some of this might change. Will be reprioritized after the release.

**Comment by gbraad** [comment]
The `sudo` usage is only on the setup phase. We explicitly disallow the use of the `root` or `administrator` to execute CRC. If administrative actions are needed, this might be performed by the admin-helper. However, this is only the case on Windows.

One of my colleagues will fill you in on more details.

----

> Creates /etc/libvirt/qemu/crc.xml
> Creates /etc/libvirt/storage/crc.xml

We do not consider these changes performed by CRC. We ask qemu+kvm to create a VM and the daemon of the virtualization stack creates these files. It depends on the target distribution where these files are created or perhaps even additional changes are made.

> Creates /etc/libvirt/qemu/networks/autostart/crc.xml
> Creates /etc/libvirt/qemu/networks/crc.xml

With this we create our own bridge to prevent overlap or conflicts with the user specified network setup. This is only the case when 'system mode networking' is used.

----

It might help to also have a look at the preflight checks we perform. These checks determine if the system is ready for use of CRC, and when not, the preflight has fixes that make these changes as aforementioned.

  * https://github.com/code-ready/crc/tree/master/pkg/crc/preflight

**Comment by gbraad** [comment]
> Changes /etc/resolv.conf (and see also https://github.com/code-ready/crc/issues/1663 which I ran into)

This should have been resolved as we support setups with `systemd-resolved`. If not, this needs further investigation and more information to explain what the situation is and what you did to resolve this.

**Comment by cfergeau** [comment]
Networking setup is indeed more complex than that:

### If using system networking
#### If systemd-resolved is used on the system
- `/etc/NetworkManager/dispatcher.d/99-crc.sh` is created

#### If systemd-resolved is not used
- Creates `/etc/NetworkManager/conf.d/crc-nm-dnsmasq.conf`
- Creates `/etc/NetworkManager/dnsmasq.d/crc.conf`
- `/etc/resolv.conf` is **not** modified by crc, this is a consequence of the configuration changes above, and an implementation detail of NetworkManager dnsmasq support.

### If using usermode networking
- `/etc/hosts` will be modified by a suid helper

If I remember correctly `/etc/hosts` is also modified in system networking even if it's not required there.

**Comment by cfergeau** [comment]
> The `sudo` usage is only on the setup phase. We explicitly disallow the use of the `root` or `administrator` to execute CRC. If administrative actions are needed, this might be performed by the admin-helper. However, this is only the case on Windows.
> 
> One of my colleagues will fill you in on more details.
> 
> > Creates /etc/libvirt/qemu/crc.xml
> > Creates /etc/libvirt/storage/crc.xml
> 
> We do not consider these changes performed by CRC. We ask qemu+kvm to create a VM and the daemon of the virtualization stack creates these files. It depends on the target distribution where these files are created or perhaps even additional changes are made.
> 
> > Creates /etc/libvirt/qemu/networks/autostart/crc.xml
> > Creates /etc/libvirt/qemu/networks/crc.xml
> 
> With this we create our own bridge to prevent overlap or conflicts with the user specified network setup. This is only the case when 'system mode networking' is used.

`crc` on linux uses libvirt to manage its VM, and it creates a `crc` storage pool, a `crc` network, and a `crc` domain in `qemu:///system`. I agree how libvirtd stores this on the filesystem is an internal libvirtd implementation detail.
```
$ virsh -c qemu:///system list --all
$ virsh -c qemu:///system pool-list --all
$ virsh -c qemu:///system net-list --all
```

**Comment by dletai** [comment]
Can the crc setup code be modified to not require sudo? I have a user with libvirt group which can create and manage VMs just fine without sudo, yet crc fails setup due to not having sudo privileges. 
```
crc setup
INFO Using bundle path /home/libvirt-user/.crc/cache/crc_okd_libvirt_4.14.0-0.okd-2023-12-01-225814_amd64.crcbundle 
INFO Checking if running as non-root              
INFO Checking if running inside WSL2              
INFO Checking if crc-admin-helper executable is cached 
INFO Caching crc-admin-helper executable          
INFO Using root access: Changing ownership of /home/libvirt-user/.crc/bin/crc-admin-helper-linux 
[sudo] password for libvirt-user:
```
Why would you require a file in a user's home dir to not be owned by the user? That just seems wrong.

**Reply by cfergeau** [reply]
It needs this file to be suid root to be able to modify /etc/hosts.
You can look at `crc config` output for the list of available configuration keys, the ones named `skip-xxx` can be set to skip most of these checks if they are not needed for you.

**Reply by dletai** [reply]
Thanks!
I'm not sure how this would mesh with your vision, but you could wrap it partially in package management, E.g., rpm, then put all such suid files in system paths, and just have crc setup configure user specific stuff. Would that work?
